Flaky Test – `integration/admin.cy.js` Timeout

This test has started failing in the last week's production builds. We should address the timeout in the test if it's reasonable, but should also consider what the bottlenecks in the admin interface are if we aren't clear. I know the atlas user queries can drag out.

```
2:13:21 AM:   Running:  integration/admin.cy.js                                                        (3 of 68)
2:13:24 AM: 
2:13:24 AM:   Admin
2:13:25 AM:     ✓ Should show not enough permissions message (1247ms)
2:13:46 AM:     (Attempt 1 of 3) Should display a list of users, their roles and allow edition
2:14:01 AM:     (Attempt 2 of 3) Should display a list of users, their roles and allow edition
2:14:16 AM:     1) Should display a list of users, their roles and allow edition
2:14:21 AM:     ✓ Should display New Incident button (4740ms)
2:14:29 AM:     ✓ Should filter results (7027ms)
2:14:29 AM:   3 passing (1m)
2:14:29 AM:   1 failing
2:14:29 AM:   1) Admin
2:14:29 AM:        Should display a list of users, their roles and allow edition:
2:14:29 AM:      AssertionError: Timed out retrying after 4000ms: Expected to find element: `[data-cy="input-filter-Id"]`, but never found it.
2:14:29 AM:       at Context.eval (webpack://gatsby-gitbook-aiid/./cypress/e2e/integration/admin.cy.js:39:13)
2:14:29 AM: 
2:14:29 AM:   (Results)
2:14:29 AM:   
```

--- a/site/gatsby-site/cypress/e2e/integration/admin.cy.js
+++ b/site/gatsby-site/cypress/e2e/integration/admin.cy.js
@@ -65,7 +65,7 @@ describe('Admin', () => {
         }
       }
 
-      const user = users.data.users.find((user) => user.userId == '63320ce63ec803072c9f529c');
+      const [user] = users.data.users;
 
       cy.get('[data-cy="input-filter-Id"]').clear();
       cy.get('[data-cy="input-filter-Id"]').type(user.userId);

--- a/site/gatsby-site/cypress/e2e/integration/admin.cy.js
+++ b/site/gatsby-site/cypress/e2e/integration/admin.cy.js
@@ -1,6 +1,5 @@
 import { conditionalIt } from '../../support/utils';
 import users from '../../fixtures/users/users.json';
-const { gql } = require('@apollo/client');
 
 describe('Admin', () => {
   const baseUrl = '/admin';
@@ -17,149 +16,152 @@ describe('Admin', () => {
     () => {
       cy.login(Cypress.env('e2eUsername'), Cypress.env('e2ePassword'));
 
+      cy.conditionalIntercept(
+        '**/graphql',
+        (req) => req.body.operationName == 'FindUsers',
+        'findUsers',
+        users
+      );
+
+      users.data.users.forEach((user, index) => {
+        cy.conditionalIntercept(
+          '**/graphql',
+          (req) =>
+            req.body.operationName == 'FindUser' &&
+            req.body?.variables?.query.userId == user.userId,
+          'findUser' + index,
+          { data: { user } }
+        );
+      });
+
       cy.visit(baseUrl);
 
-      cy.query({
-        query: gql`
-          {
-            users {
-              userId
-              roles
-              adminData {
-                email
-              }
-            }
-          }
-        `,
-        timeout: 60000, // mongodb admin api is extremely slow
-      }).then(({ data: { users } }) => {
-        cy.waitForStableDOM();
+      cy.waitForStableDOM();
 
-        for (const user of users.slice(0, 5)) {
-          cy.get('[data-cy="input-filter-Id"]').clear();
-          cy.get('[data-cy="input-filter-Id"]').type(user.userId);
-
-          if (user.adminData.email) {
-            cy.contains('[data-cy="cell"]', user.userId)
-              .parent()
-              .within(() => {
-                cy.contains(user.adminData.email).scrollIntoView().should('be.visible');
-
-                for (const role of user.roles) {
-                  cy.contains(role).scrollIntoView().should('be.visible');
-                }
-              });
-          } else {
-            cy.contains('[data-cy="cell"]', user.userId)
-              .parent()
-              .within(() => {
-                cy.contains('Not found').scrollIntoView().should('be.visible');
-
-                for (const role of user.roles) {
-                  cy.contains(role).scrollIntoView().should('be.visible');
-                }
-              });
-          }
-        }
-
-        const user = users.find((user) => user.adminData.email == Cypress.env('e2eUsername'));
-
+      for (const user of users.data.users.slice(0, 5)) {
         cy.get('[data-cy="input-filter-Id"]').clear();
         cy.get('[data-cy="input-filter-Id"]').type(user.userId);
 
-        cy.contains('[data-cy="cell"]', user.adminData.email)
-          .parent()
-          .within(() => {
-            cy.contains('Edit').click();
-          });
+        if (user.adminData.email) {
+          cy.contains('[data-cy="cell"]', user.userId)
+            .parent()
+            .within(() => {
+              cy.contains(user.adminData.email).scrollIntoView().should('be.visible');
 
-        cy.waitForStableDOM();
+              for (const role of user.roles) {
+                cy.contains(role).scrollIntoView().should('be.visible');
+              }
+            });
+        } else {
+          cy.contains('[data-cy="cell"]', user.userId)
+            .parent()
+            .within(() => {
+              cy.contains('Not found').scrollIntoView().should('be.visible');
 
-        cy.conditionalIntercept(
-          '**/graphql',
-          (req) => req.body.operationName == 'UpdateUserRoles',
-          'UpdateUserRoles',
-          {
-            data: {
-              updateOneUser: {
+              for (const role of user.roles) {
+                cy.contains(role).scrollIntoView().should('be.visible');
+              }
+            });
+        }
+      }
+
+      const user = users.data.users.find((user) => user.userId == '63320ce63ec803072c9f529c');
+
+      cy.get('[data-cy="input-filter-Id"]').clear();
+      cy.get('[data-cy="input-filter-Id"]').type(user.userId);
+
+      cy.contains('[data-cy="cell"]', user.adminData.email)
+        .parent()
+        .within(() => {
+          cy.contains('Edit').click();
+        });
+
+      cy.waitForStableDOM();
+
+      cy.conditionalIntercept(
+        '**/graphql',
+        (req) => req.body.operationName == 'UpdateUserRoles',
+        'UpdateUserRoles',
+        {
+          data: {
+            updateOneUser: {
+              __typename: 'User',
+              roles: ['subscriber', 'bananas', 'ban', 'admin', 'banana'],
+              userId: '6423479655e4bb918a233bda',
+            },
+          },
+        }
+      );
+
+      cy.conditionalIntercept(
+        '**/graphql',
+        (req) => req.body.operationName == 'UpdateUserProfile',
+        'UpdateUserProfile',
+        {
+          data: {
+            users: [
+              {
                 __typename: 'User',
-                roles: ['subscriber', 'bananas', 'ban', 'admin', 'banana'],
+                adminData: {
+                  __typename: 'UserAdminDatum',
+                  creationDate: '2023-03-28T20:01:26Z',
+                  disabled: false,
+                  email: 'mail@cesarvarela.com',
+                  lastAuthenticationDate: '2023-05-04T22:03:06Z',
+                },
+                first_name: 'TestEdited',
+                last_name: 'User',
+                roles: ['subscriber', 'admin', 'banana'],
                 userId: '6423479655e4bb918a233bda',
               },
-            },
-          }
-        );
-
-        cy.conditionalIntercept(
-          '**/graphql',
-          (req) => req.body.operationName == 'UpdateUserProfile',
-          'UpdateUserProfile',
-          {
-            data: {
-              users: [
-                {
-                  __typename: 'User',
-                  adminData: {
-                    __typename: 'UserAdminDatum',
-                    creationDate: '2023-03-28T20:01:26Z',
-                    disabled: false,
-                    email: 'mail@cesarvarela.com',
-                    lastAuthenticationDate: '2023-05-04T22:03:06Z',
-                  },
-                  first_name: 'TestEdited',
-                  last_name: 'User',
-                  roles: ['subscriber', 'admin', 'banana'],
-                  userId: '6423479655e4bb918a233bda',
+              {
+                __typename: 'User',
+                adminData: {
+                  __typename: 'UserAdminDatum',
+                  creationDate: '2023-03-30T19:04:55Z',
+                  disabled: false,
+                  email: 'aiidsubscriber@cesarvarela.com',
+                  lastAuthenticationDate: '2023-04-28T00:33:58Z',
                 },
-                {
-                  __typename: 'User',
-                  adminData: {
-                    __typename: 'UserAdminDatum',
-                    creationDate: '2023-03-30T19:04:55Z',
-                    disabled: false,
-                    email: 'aiidsubscriber@cesarvarela.com',
-                    lastAuthenticationDate: '2023-04-28T00:33:58Z',
-                  },
-                  first_name: 'cesarito',
-                  last_name: 'test',
-                  roles: ['subscriber', 'incident_editor'],
-                  userId: '6425dd57ea44a80bf1443b33',
+                first_name: 'cesarito',
+                last_name: 'test',
+                roles: ['subscriber', 'incident_editor'],
+                userId: '6425dd57ea44a80bf1443b33',
+              },
+              {
+                __typename: 'User',
+                adminData: {
+                  __typename: 'UserAdminDatum',
+                  creationDate: null,
+                  disabled: null,
+                  email: null,
+                  lastAuthenticationDate: null,
                 },
-                {
-                  __typename: 'User',
-                  adminData: {
-                    __typename: 'UserAdminDatum',
-                    creationDate: null,
-                    disabled: null,
-                    email: null,
-                    lastAuthenticationDate: null,
-                  },
-                  first_name: null,
-                  last_name: null,
-                  roles: ['subscriber', 'admin'],
-                  userId: '6425dd79c505cc40eb65661e',
-                },
-              ],
-            },
-          }
-        );
+                first_name: null,
+                last_name: null,
+                roles: ['subscriber', 'admin'],
+                userId: '6425dd79c505cc40eb65661e',
+              },
+            ],
+          },
+        }
+      );
 
-        cy.get('[data-cy="edit-user-modal"]').within(() => {
-          cy.get('[id="roles"]', { timeout: 30000 }).type('banana{enter}');
+      cy.get('[data-cy="edit-user-modal"]').within(() => {
+        cy.get('[id="roles"]', { timeout: 30000 }).type('banana{enter}');
 
-          cy.get('[name="first_name"]').clear().type('Edited');
+        cy.get('[name="first_name"]').clear().type('Edited');
 
-          cy.contains('Submit').click();
+        cy.contains('Submit').click();
 
-          cy.wait('@UpdateUserRoles').then((xhr) => {
-            expect(xhr.request.body.variables.roles).includes('banana');
-            expect(xhr.request.body.variables.userId).eq(user.userId);
-          });
+        cy.wait('@UpdateUserRoles').then((xhr) => {
+          expect(xhr.request.body.variables.roles).includes('banana');
+          expect(xhr.request.body.variables.userId).eq(user.userId);
+        });
 
-          cy.wait('@UpdateUserProfile').then((xhr) => {
-            expect(xhr.request.body.variables.first_name).eq('Edited');
-            expect(xhr.request.body.variables.userId).eq(user.userId);
-          });
+        cy.wait('@UpdateUserProfile').then((xhr) => {
+          expect(xhr.request.body.variables.first_name).eq('Edited');
+          expect(xhr.request.body.variables.userId).eq(user.userId);
         });
       });
     }

--- a/site/gatsby-site/cypress/fixtures/users/users.json
+++ b/site/gatsby-site/cypress/fixtures/users/users.json
@@ -80,6 +80,22 @@
           "subscriber"
         ],
         "userId": "619b47ea5eed5334edfa3bbc"
+      },
+      {
+        "__typename": "User",
+        "adminData": {
+          "__typename": "UserAdminDatum",
+          "creationDate": null,
+          "disabled": null,
+          "email": null,
+          "lastAuthenticationDate": null
+        },
+        "first_name": null,
+        "last_name": null,
+        "roles": [
+          "subscriber"
+        ],
+        "userId": "63e5d7a329e0d8b04a5092a6"
       }
     ]
   }

